### PR TITLE
Fix `interval` to use consistent units and arrow parser

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/5801

# Rationale for this change

1. To consistently support intervals we need to uniformly handle that type
2. @doki23 ported the interval parsing code uptream in https://github.com/apache/arrow-rs/pull/3762 so we should use that (in preparation of supporting intervals more fully in DataFusion)

# What changes are included in this PR?

1. Remove datafusion special `parse_interval` function and use arrow kernel everyhere

# Are these changes tested?
yes, both existing and new tests

# Are there any user-facing changes?
More consistency with how `interval ` is handled

### CODE REVIEW

File: datafusion/sql/Cargo.toml [hunk 1]
```diff
@@ -38,6 +38,7 @@ unicode_expressions = []
 
 [dependencies]
 arrow-schema = { workspace = true }
+arrow = { workspace = true }
 datafusion-common = { path = "../common", version = "21.0.0" }
```
Comment (alamb) [inline datafusion/sql/Cargo.toml:42]: `datafusion-common` already depends on arrow so this is no new (net) dependency

File: datafusion/sql/src/expr/value.rs [hunk 2]
```diff
@@ -202,11 +203,17 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-            .unwrap_or_else(|| "second".to_string());
+        // leading_field really means the unit if specified
+        // for example, "month" in  `INTERVAL '5' month`
 
-        Ok(lit(parse_interval(&leading_field, &value)?))
+        let value = if let Some(leading_field) = leading_field.as_ref() {
+            format!("{value} {leading_field}")
+        } else {
+            // default to seconds for the units
+            format!("{value} seconds")
```
Comment (alamb) [inline datafusion/sql/src/expr/value.rs:213]: this is the same default as previously

File: datafusion/optimizer/src/type_coercion.rs [hunk 3]
```diff
@@ -601,14 +593,7 @@ fn coerce_arguments_for_signature(
 
 /// Cast `expr` to the specified type, if possible
 fn cast_expr(expr: &Expr, to_type: &DataType, schema: &DFSchema) -> Result<Expr> {
-    // Special case until Interval coercion is handled in arrow-rs
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:604]: 🎉  for avoiding the special cases

File: datafusion/common/src/parsers.rs [hunk 4]
```diff
@@ -76,337 +72,3 @@ impl CompressionTypeVariant {
         !matches!(self, &Self::UNCOMPRESSED)
     }
 }
-
```
Comment (alamb) [inline datafusion/common/src/parsers.rs:79]: The point of this PR is to remove this code (which has been upstreamed to arrow-rs)

File: datafusion/core/tests/sql/expr.rs [hunk 5]
```diff
@@ -843,153 +843,153 @@ async fn test_interval_expressions() -> Result<()> {
...
     );
     test_expression!(
         "interval '5 month'",
-        "0 years 5 mons 0 days 0 hours 0 mins 0.00 secs"
+        "0 years 5 mons 0 days 0 hours 0 mins 0.000000000 secs"
     );
     test_expression!(
         "interval '13 month'",
-        "1 years 1 mons 0 days 0 hours 0 mins 0.00 secs"
+        "0 years 13 mons 0 days 0 hours 0 mins 0.000000000 secs"
```
Comment (alamb) [inline datafusion/core/tests/sql/expr.rs:964]: I am not sure this matters -- but `interval` parsing now is `month/day/nano` always and thus is not quite as nice a display as it was previously (no years, instead it is always months). However perhaps if anyone cares we can improve the display in arrow-rs for this kind of interval

File: datafusion/core/tests/sql/timestamp.rs [hunk 6]
```diff
@@ -968,52 +968,52 @@ async fn timestamp_array_add_interval() -> Result<()> {
...
     let expected = vec![
-        "+----------------------------+--------------------------------------+",
-        "| ts                         | table_b.ts + IntervalDayTime(\"1000\") |",
-        "+----------------------------+--------------------------------------+",
-        "| 2020-09-08T13:42:29.190855 | 2020-09-08T13:42:30.190855           |",
-        "| 2020-09-08T12:42:29.190855 | 2020-09-08T12:42:30.190855           |",
-        "| 2020-09-08T11:42:29.190855 | 2020-09-08T11:42:30.190855           |",
-        "+----------------------------+--------------------------------------+",
+        "+----------------------------+-------------------------------------------------+",
+        "| ts                         | table_b.ts + IntervalMonthDayNano(\"1000000000\") |",
```
Comment (alamb) [inline datafusion/core/tests/sql/timestamp.rs:985]: This is pretty terrible formatting for the column name 🤮  but I don't think it is fixable easily -- see #2027 for more discussion

File: datafusion/sql/src/expr/value.rs [hunk 7]
```diff
@@ -202,11 +203,55 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             }
         };
 
-        let leading_field = leading_field
-            .as_ref()
-            .map(|dt| dt.to_string())
-            .unwrap_or_else(|| "second".to_string());
+        let value = if has_units(&value) {
```
Comment (alamb) [inline datafusion/sql/src/expr/value.rs:206]: This is a pretty terrible hack but it seems to be necessary to get the tests to pass (aka to have postgres compatible syntax) 🤷

File: datafusion/core/tests/sqllogictests/test_files/interval.slt [hunk 8]
```diff
@@ -23,7 +23,113 @@ select
   arrow_typeof(interval '5 months'),
   arrow_typeof(interval '5 days 3 nanoseconds')
 ----
-Interval(YearMonth) Interval(MonthDayNano)
+Interval(MonthDayNano) Interval(MonthDayNano)
+
+
+## This is incredibly confusing but document it in tests:
+#
+# years is parsed as a column name
```
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/interval.slt:31]: Is this a bug?

**Review by tustvold — APPROVED** [review]

File: datafusion/core/tests/sqllogictests/test_files/interval.slt (hunk 8 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/interval.slt:31]: DataFusion is consistent with postgres so in that sense it is not a bug 

This behavior is due to how the statement is parsed (sqlparser-rs treats `years` as the alias)

It is incredibly confusing (I was confused for a while)

Now that we have proper `Utf8` -> `Interval` parsing (thanks to @doki23 ) in arrow-rs I think the preferred way to create intervals should be `'5 months'::interval` which does the right thing and doesn't have the wacky syntax

File: datafusion/core/tests/sql/expr.rs (hunk 5 above)
Comment (alamb) [reply datafusion/core/tests/sql/expr.rs:964]: @waitingkuo  and @ovr  -- Do you have any thoughts about this proposed change (that the reported intervals are not in the same units as the original expression)?

Comment (tustvold) [reply datafusion/core/tests/sql/expr.rs:964]: FWIW this was a conscious change when the parsing logic was ported, to faithfully preserve what the user specified. This was mainly in the context of rounding days to months, which was simply incorrect, but months to years got the same treatment

File: datafusion/sql/src/expr/value.rs (hunk 7 above)
Comment (tustvold) [reply datafusion/sql/src/expr/value.rs:206]: Would it be valid to just search for a character that isn't a digit or decimal point?

**Review by jackwener — APPROVED** [review]
A great job! Thanks @alamb

File: datafusion/core/tests/sqllogictests/test_files/interval.slt [hunk 9]
```diff
@@ -23,7 +23,113 @@ select
...
+select interval '5' years
+----
+0 years 0 mons 0 days 0 hours 0 mins 5.000000000 secs
+
+
+# check all different kinds of intervals
+query ?
+select interval '5' year
+----
+0 years 60 mons 0 days 0 hours 0 mins 0.000000000 secs
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/interval.slt:55]: Yeah, I agree it is a little strange. I can't figure out how important it is to to mirror back exactly what the user said.

For some things like `5 years` we could pick `Interval(YearMonth)` and encode exactly what the user specified

For things like `5 years 3 days` to express that properly I think we need to use `Interval(MonthDayNano)` which is going to be formatted like `60 months ... 3 days`

Or maybe we should just make the interval printing prettier 🤔  like convert `60 months` to 5 years...

File: datafusion/core/tests/sql/expr.rs (hunk 5 above)
Comment (jacksonrnewhouse) [reply datafusion/core/tests/sql/expr.rs:964]: Exploring how postgres handles this, if you do EXTRACT(EPOCH FROM interval), it treats months as 30 days, until you get to 12 months, and then it is 365.25 days. Not sure how much it matters to match postgres, but that's what I've seen.

Comment (alamb) [reply datafusion/core/tests/sql/expr.rs:964]: 🤯

**Pull request merged**